mobile/src/main/java/org/openhab/habdroid/core/connection/ConnectionFactory.java

```diff
@@ -0,0 +1,282 @@
+                    ConnectivityManager.TYPE_WIMAX));
+    private static final List<String> needInvalidateCachePreferenceKeys = Arrays.asList(Constants
+            .PREFERENCE_ALTURL, Constants.PREFERENCE_LOCAL_USERNAME, Constants
+            .PREFERENCE_LOCAL_PASSWORD, Constants.PREFERENCE_REMOTE_PASSWORD, Constants
+            .PREFERENCE_REMOTE_USERNAME, Constants.PREFERENCE_URL, Constants.PREFERENCE_DEMOMODE);
+
+    private Context ctx;
+    private SharedPreferences settings;
+    private NetworkInfo activeNetworkInfo;
+
+    Map<Integer, Connection> cachedConnections = new HashMap<>();
+
+    private static class InstanceHolder {
+        public static final ConnectionFactory INSTANCE = new ConnectionFactory();
+    }
+
+    static ConnectionFactory getInstance() {
+        return InstanceHolder.INSTANCE;
+    }
+
+    public ConnectionFactory() {
+        if (Build.VERSION.SDK_INT >= 21) {
+            localConnectionTypes.add(ConnectivityManager.TYPE_VPN);
+        }
+    }
+
+    private void setContext(Context ctx) {
+        this.ctx = ctx;
+    }
+
+    @Override
+    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
+        if (needInvalidateCachePreferenceKeys.contains(key)) {
+            cachedConnections.clear();
+        }
+    }
+
+    private void setSettings(SharedPreferences settings) {
+        if (this.settings != null) {
+            this.settings.unregisterOnSharedPreferenceChangeListener(this);
+        }
+        this.settings = settings;
+
+        this.settings.registerOnSharedPreferenceChangeListener(this);
+    }
+
+    public static Connection getConnection(int connectionType, Context ctx) {
+        return getConnection(connectionType, ctx,
+                PreferenceManager.getDefaultSharedPreferences(ctx));
+    }
+
+    public static Connection getConnection(int connectionType, Context ctx,
+                                           SharedPreferences settings) {
+        // FIXME: Once we're on API level 19, use Objects.requireNonNull();
+        if (ctx == null) {
+            throw new NullPointerException("Context ctx can not be null.");
+        }
+
+        ConnectionFactory factory = getInstance();
+        factory.setSettings(settings);
+        factory.setContext(ctx);
+
+        ctx.registerReceiver(factory, new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION));
+
+        if (factory.cachedConnections.containsKey(Connection.TYPE_ANY)) {
+            return factory.cachedConnections.get(Connection.TYPE_ANY);
+        }
+
+        Connection conn;
+        switch (connectionType) {
+            case Connection.TYPE_LOCAL:
+                conn = factory.getLocalConnection();
+                break;
+            case Connection.TYPE_REMOTE:
+                conn = factory.getRemoteConnection();
+                break;
+            case Connection.TYPE_CLOUD:
+                // TODO: Need a proper way of finding if the connection supports openHAB cloud
```

Comment:
Why doesnt it work on myopenhab.org? Isnt it better to hardcode it until it is fixed, instead of having "Notifications" visible for everyone?

Reply:
Because myopenhab.org has an outdated version of openhab-cloud, it would need at least commit https://github.com/openhab/openhab-cloud/commit/f7e1aaf6f3bc2ffd295c1addca85284aee1f6b66 to have the settings api endpoint.

I wouldn't like to hardcode it, as it would be a step back, as third-party hosted openhab-cloud instances can't be used anymore :( For now, it shouldn't have changed to the current version of the app, which shows the Notifications tab, once the remote URL is set, isn't it?

Reply:
Its always shown.

Reply:
> Because myopenhab.org has an outdated version of openhab-cloud

Note that @marziman & @digitaldan were just about to deploy an updated version - maybe you want to wait for it, so that no workarounds are necessary.

Reply:
> Its always shown.

Not with the last update, anymore. There was an error in the connection caching, which results in that behaviour, sorry for the confusion :)

> Note that @marziman & @digitaldan were just about to deploy an updated version - maybe you want to wait for it, so that no workarounds are necessary.

Oh, great! Thanks for the info. I'm not sure, if we _need_ to wait here. The behaviour doesn't change with this PR (if remote URL is set, notifications are shown, no matter, if the remote URL points to an openHAB-cloud instance). However, implementing that it is only shown for openHAB-cloud instances is something bigger, e.g., I wouldn't like to do that check in the ConnectionFactory, as it would require to be a sync http call (no caller expects to wait for an async call when trrying to get a connection object), which would mean the call is dependant upon the response time of the remote URL. So, I would do that when the remote url is changed, having a progress dialog saying, that the url is checked and, if it's an openHAB-cloud instance, set a specific unvisible preference, which is used in here.

But, like I said: Having such a functionality is something I want to have, too, however, it would make this PR even bigger and bigger (which is already quite big). And as long, as the behaviour does not signigicantly change in this PR, I would suggest to implement this improvement in a separate PR :]